### PR TITLE
Basic authentication is done

### PR DESCRIPTION
**Pull request by ajcpwnz**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  def new
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user && user.authenticate(params[:session][:password])
+      sign_in user
+      redirect_to user
+    else
+      render 'new'
```
Comment (IlyaUmanets) [inline app/controllers/sessions_controller.rb]: `render :new`

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    end
+  end
+
+  def destroy
+    sign_out
+    redirect_to root_url
+  end
+end
```
Comment (IlyaUmanets) [inline app/controllers/sessions_controller.rb]: Add new line please.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params) 
+    if @user.save
+      sign_in @user
+      redirect_to @user
+    else
+      render 'new'
```
Comment (IlyaUmanets) [inline app/controllers/users_controller.rb]: Same

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  private
+
+    def user_params
+      params.require(:user).permit(:username, :email, :password,
+                                   :password_confirmation)
+    end
+
+end
```
Comment (IlyaUmanets) [inline app/controllers/users_controller.rb]: Same

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    @current_user ||= User.find_by(remember_token: remember_token)
+  end
+
+  def sign_out
+    current_user.update_attribute(:remember_token,
+                                  User.encrypt(User.new_remember_token))
+    cookies.delete(:remember_token)
+    self.current_user = nil
+  end
+end
```
Comment (IlyaUmanets) [inline app/helpers/sessions_helper.rb]: Same

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+module SessionsHelper
+  def sign_in(user)
+    remember_token = User.new_remember_token
+    cookies.permanent[:remember_token] = remember_token
```
Comment (IlyaUmanets) [inline app/helpers/sessions_helper.rb]: Do we really need this ? What about `session[:user_id]` ?

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,27 @@
+class User < ApplicationRecord
+  before_save { self.email = email.downcase }
+  before_create :create_remember_token
+  validates :username, presence: true, 
+                       length: { maximum: 11 },
+                       uniqueness: true
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (IlyaUmanets) [inline app/models/user.rb]: Move to top of class

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  validates :email, presence:   true,
+                    format:     { with: VALID_EMAIL_REGEX },
+                    uniqueness: { case_sensitive: false }
+  has_secure_password
+  validates :password, length: { minimum: 6 }
```
Comment (IlyaUmanets) [inline app/models/user.rb]: Write `validates` rules in one section.
[Example](http://stackoverflow.com/questions/16959389/what-is-the-correct-order-of-associations-scopes-validations-etc-in-rails-mo)

File: app/models/user.rb (hunk 7 above)
Comment (ajcpwnz) [inline app/models/user.rb]: I propose to add validation as separate concern, what do you think?

File: app/models/user.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
+class User < ApplicationRecord
+  before_save { self.email = email.downcase }
+  before_create :create_remember_token
+  validates :username, presence: true,
```
Comment (IlyaUmanets) [inline app/models/user.rb]: Please add new line to divide `callbacks` and `validates` section.

File: app/models/user.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+                       length: { maximum: 11 },
+                       uniqueness: true
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence:   true,
+                    format:     { with: VALID_EMAIL_REGEX },
+                    uniqueness: { case_sensitive: false }
+  has_secure_password
+  validates :password, length: { minimum: 6 }
+
+  def User.new_remember_token
```
Comment (IlyaUmanets) [inline app/models/user.rb]: `def self.`. If not please explain your approach.

File: app/models/user.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    SecureRandom.urlsafe_base64
+  end
+
+  def User.encrypt(token)
```
Comment (IlyaUmanets) [inline app/models/user.rb]: Same

File: app/models/user.rb [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+    Digest::SHA1.hexdigest(token.to_s)
+  end
+
+  private
+
+    def create_remember_token
+      self.remember_token = User.encrypt(User.new_remember_token)
```
Comment (IlyaUmanets) [inline app/models/user.rb]: `self.remember_token = encrypt(new_remember_token)`

File: app/views/sessions/new.html.slim [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+  p Sign In
+  div
+    = form_for(:session, url: sessions_path) do |user|
+      p = user.label :email
+      p = user.text_field :email
+
+      p = user.label :password
+      p = user.password_field :password
+
+      p = user.submit "Submit"
```
Comment (IlyaUmanets) [inline app/views/sessions/new.html.slim]: Same

File: app/views/sessions/new.html.slim [hunk 14]
```diff
@@ -0,0 +1,11 @@
+div
+  p Sign In
+  div
+    = form_for(:session, url: sessions_path) do |user|
+      p = user.label :email
+      p = user.text_field :email
```
Comment (IlyaUmanets) [inline app/views/sessions/new.html.slim]: add `required: true` option.

File: app/views/users/new.html.slim [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+      p = user.label :email
+      p = user.text_field :email
+
+      p = user.label :password
+      p = user.password_field :password
+
+      p = user.label :password_confirmation
+      p = user.password_field :password_confirmation
+
+      p = user.submit "Submit"
```
Comment (IlyaUmanets) [inline app/views/users/new.html.slim]: Same

File: app/views/users/show.html.slim [hunk 16]
```diff
@@ -0,0 +1,7 @@
+div
+  p Profile
+  div 
+    p = @user.username
+    p = @user.email
+    - if signed_in?
+      p = link_to "Sign out", session_path, method: "delete"
```
Comment (IlyaUmanets) [inline app/views/users/show.html.slim]: Same

File: spec/factories/users.rb [hunk 17]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
```
Comment (IlyaUmanets) [inline spec/factories/users.rb]: Remove this generated file.
Capybara scenarios will be good.

**Comment by IlyaUmanets** [comment]
Specs will be good.

File: app/views/pages/stub.html.slim [hunk 18]
```diff
@@ -0,0 +1,9 @@
+.dashboard__create-team
+  h2 That's all folks.
+  p Thank you for your interest, #{current_user.username}.
+  p We've got your email and will let you know if there are any updates.
+  p
+    | Meanwhile, you can check out our
+    = link_to ' repo', t('site.misc.repo_url'), target: '_blank'
+    | .
+  p = link_to 'Sign Out', sign_out_path
```
Comment (IlyaUmanets) [inline app/views/pages/stub.html.slim]: Same

File: spec/actions/user_sessions/registration_spec.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+feature 'Sign up' do
+  let(:user) { build(:user) }
+
+  scenario 'user submits without entering details' do
+    visit sign_up_path
+    fill_in 'user[username]', with: ''
```
Comment (IlyaUmanets) [inline spec/actions/user_sessions/registration_spec.rb]: `fill_in label_text`

File: spec/actions/user_sessions/registration_spec.rb [hunk 20]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+feature 'Sign up' do
+  let(:user) { build(:user) }
```
Comment (IlyaUmanets) [inline spec/actions/user_sessions/registration_spec.rb]: `attributes_for(:user)`

File: spec/actions/user_sessions/auth_spec.rb [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+  scenario 'user acesses signin page' do
+    visit root_path
+    expect(page).to have_css 'a', text: 'Sign Up'
+    expect(page).to have_css 'a', text: 'Sign In'
+    click_link 'Sign In'
+  end
+
+  scenario 'user submits without entering details' do
+    visit sign_in_path
+    fill_in 'session[email]', with: ''
```
Comment (IlyaUmanets) [inline spec/actions/user_sessions/auth_spec.rb]: I know that you don't have labels here. But you have placeholders.
Try it
`fill_in :placeholder => "Fill in some text", :with => "Text"`

File: spec/actions/user_sessions/registration_spec.rb [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+  scenario 'user submits without entering details' do
+    visit sign_up_path
+    fill_in 'user[username]', with: ''
+    click_button 'Sign Up'
+    expect(page).to have_content 'Please fill in required fields'
+  end
+
+  scenario 'user signs in with valid details' do
+    visit sign_up_path
+    fill_in 'user[username]', with: user.username
```
Comment (IlyaUmanets) [inline spec/actions/user_sessions/registration_spec.rb]: You have section with same actions. In this case you call `fill_in` 3 times. Please put empty line before and afer this section. It's more readable

File: spec/actions/user_sessions/auth_spec.rb [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+  scenario 'user signs in with valid details and signs out' do
+    visit sign_in_path
+    fill_in 'session[email]', with: user.email
+    fill_in 'session[password]', with: user.password
+    click_button 'Sign In'
+    expect(page).to have_content user.username
+    expect(page).to have_content 'Sign Out'
+    # current flow does not allow to test signout from anywhere
+    # TODO test signout scenario when real dashboard is ready
+    click_link 'Sign Out'
```
Comment (IlyaUmanets) [inline spec/actions/user_sessions/auth_spec.rb]: You have to move this to seperate scenario. This trick will help you.
`allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(user)`

**Pull request merged**